services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/ProjectSettingsManagerImpl.java

```diff
@@ -242,6 +251,92 @@ public StorageLocationSetting getStorageLocationSetting(Long storageLocationId)
 		}
 		return storageLocationDAO.get(storageLocationId);
 	}
+	
+	@Override
+	public MergeStorageLocationsResponse mergeDuplicateStorageLocations(UserInfo userInfo) throws DatastoreException, UnauthorizedException {
+		ValidateArgument.required(userInfo, "userInfo");
+		
+		if (!userInfo.isAdmin()) {
+			throw new UnauthorizedException("Only administrators can invoke this service");
+		}
+		
+		// Maps a duplicate storage location id to the id of the storage location we keep
+		Map<Long, Long> duplicatesMap = new HashMap<>();
+		
+		for (Long id : storageLocationDAO.findAllWithDuplicates()) {
+			// Already processed within another set
+			if (duplicatesMap.containsKey(id)) {
+				continue;
+			}
+			
+			Set<Long> duplicates = storageLocationDAO.findDuplicates(id);
+	
+			// Updates all the file handles
+			if (!duplicates.isEmpty()) {
+				fileHandleDao.updateStorageLocationBatch(duplicates, id);
+			}
+			
+			duplicates.forEach(duplicateId  -> duplicatesMap.put(duplicateId, id));
+		}
+		
+		Long updatedProjectsCount = 0l;
+
+		if (!duplicatesMap.isEmpty()) {
+			// Updates the projects referencing the duplicate ids
+			updatedProjectsCount = removeDuplicateStorageLocationsFromProjects(duplicatesMap);
+			
+			// Finally drop the unused storage location settings
+			storageLocationDAO.deleteBatch(duplicatesMap.keySet());
+		}
+		
+		MergeStorageLocationsResponse response = new MergeStorageLocationsResponse();
+		
+		response.setDuplicateLocationsCount((long) duplicatesMap.size());
+		response.setUpdatedProjectsCount(updatedProjectsCount);
+
+		return response;
+	}
+	
+	Long removeDuplicateStorageLocationsFromProjects(Map<Long, Long> duplicatesMap) {
+		Long updatedProjectsCount = 0l;
+		
+		List<ProjectSetting> projectSettings = projectSettingsDao.getByType(ProjectSettingsType.upload);
```

Comment:
- Remove the getByType after the PR.
- Check if an iterator pattern is a better option